### ISSUE TITLE
The second time it's used, deprecatedModuleAttribute still warns twice

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4806]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#4806|
|Type|defect|
|Created|2011-01-15 01:27:41Z|
|Branch|https://github.com/twisted/twisted/tree/double-deprecate-4806|

As discovered in the implementation of [#4298](https://github.com/twisted/twisted/issues/4298), one [#4492](https://github.com/twisted/twisted/issues/4492) was not quite completely fixed.

If you deprecate the attribute `module` on the package `package`, the second time you `from package import module`, you'll get 2 warnings instead of one.

Attachments:

* [deprecate-twice.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/001/001598f071bf265d2412050edfbd501eed19314b/34236ececbdc7f81fd6b36b9fddfabfd1757911a.diff) (4532 bytes) - added by glyph on 2011-01-15 01:28:04Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4806 4806
type__defect defect
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_double_deprecate_4806 branches/double-deprecate-4806
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1295054861000000 1295054861000000
changetime__1310537474000000 1310537474000000
version__None None
owner__glyph glyph

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

The attached patch demonstrates the issue - it's a test patch only, no implementation fix yet.  I also took the opportunity to fix the fact that `trial -u twisted.test.test_deprecate` fails on its second run.  (Ironically, failing with the exact issue that this ticket describes, because the test wasn't cleaning up after itself properly.)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

The reason this happens is actually obliquely explained already in the comment in `_DeprecatedAttribute.get`.  The first time that the import system comes around to see if the attribute exists, I guess in the C equivalent of a `hasattr`, to see if the name already exists, it gets an `AttributeError` and doesn't produce a warning.  But in all subsequent calls, when that check succeeds, the call to actually get the name and give it back to user code is a _separate_ `__getattribute__` invocation, and both produce a warning.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

(In [31744]) Branching to 'double-deprecate-4806'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/double-deprecate-4806), when they're ready.

The solution in the branch is only heuristically correct, but I figured it wasn't worth going to the trouble of being more correct, since you'd have to do some awful stuff (lots of stack-walking, making it brittle and even less friendly to import hooks).  It should reliably work in the default case though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @glyph|
|-|-|

Thanks.  Here are some rather nit-picky comments:

  1. The diff was rather difficult to read since it smashed the `_InternalState` change together with actual logic changes (which are somewhat subtle).  It would have been better if this could have been done as two patches, one to just switch to `_InternalState`, and one on top of that which fixed the bug described in the summary.
  1. Maybe `_ModuleProxy.__getattribute__` would be made easier to read by the addition of a comment or two explaining the difference between `deprecatedAttribute` being `None` and otherwise.
  1. Two other unit tests in `twisted.python.test.test_deprecate` still fail when run more than once; this might merit a new ticket.
  1. It might be nice to assert something about the contents of `emitted` in `checkOneWarning`, in addition to the length.  Or maybe this is redundant with existing tests (but they don't exercise this case !)
  1. It would be nice for `PySpaceTestCase` to move to a non-`test_` module and to get a class docstring.  This should make it more clear that it's a piece of re-used test infrastructure, not something particular to the `test_modules.py` suite alone.  `test_deprecate.py` will actually be the third consumer of this API (counting `test_modules.py`) now.

Of these points, I am mostly interested in the last, as a matter of improving the factoring and organization of our test suite.  Please merge when you are satisfied with the branch with respect to these issues.  Also, you might be pleased to know that this works on PyPy.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

(In [32319]) re [#4806](https://github.com/twisted/twisted/issues/4806) point 4, it is annoying to compute the whole thing (lineno, filename) but at least make sure the emitted list has the vaguest shape of a deprecation warning.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

(In [32320]) re [#4806](https://github.com/twisted/twisted/issues/4806) review point 5: move PySpaceTestCase to a new home in twisted.python.test.modules_helpers since it is a fixture and not itself a test case, and give it a less ridiculous and obsolete name (TwistedModulesTestCase)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

To point 1: in the future, I will (as always) endeavor to make my changesets shorter and clearer.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

(In [32324]) re [#4806](https://github.com/twisted/twisted/issues/4806) point 2, I think this is a bit redundant, but this code _is_ quite hard to follow, so I'll try to make it as clear as possible.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

Once more, [with feeling](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/double-deprecate-4806).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph commented|
|-|-|

For point 3, I filed [#5197](https://github.com/twisted/twisted/issues/5197).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph set status to `closed`|
|-|-|

(In [32325]) Merge double-deprecate-4806: fix double-warning about module deprecations.

Author: glyph

Reviewer: exarkun

Fixes: [#4806](https://github.com/twisted/twisted/issues/4806)

twisted.python.deprecate.deprecatedModuleAttribute no longer spuriously warns
twice when used to deprecate a module within a package.  This should make it
easier to write unit tests for deprecated modules.
